testhelpers/docker_registry.go

```diff
@@ -68,17 +68,25 @@ func NewDockerRegistry(ops ...RegistryOption) *DockerRegistry {
 }
 
 // BasicAuth wraps a handler, allowing requests with matching username and password headers, otherwise rejecting with a 401
-func BasicAuth(handler http.Handler, username, password, realm string) http.Handler {
+// optPermissions is used to skip the authentication check if the image has been
+// configured in the test setup to be always readable.
+func BasicAuth(handler http.Handler, username, password, realm string, optPermissions ...map[string]ImagePrivileges) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if len(optPermissions) > 0 && (r.Method == "GET" || r.Method == "HEAD") {
+			permissions := optPermissions[0]
+			if permission, ok := permissions[extractImageName(r.URL.Path)]; ok {
+				if permission.readable {
+					handler.ServeHTTP(w, r)
+				}
+			}
+		}
```

Comment:
This sort of bothers me in that we're duplicating logic in `Custom()` (and also not handling the case when we want to "pass" a writeable image). It seems that what we want to do is IF the url is found in the map, just fall through to the regular handler (if permission is not denied). What if instead, we send `r.regHandler` into `Custom()` - and then around line 124, instead of falling through to the auth handler, we invoke `regHandler.Serve()`. Then we could delete this logic.

Reply:
Keeping my idea of expressing the behavior with a diagram, doing this will work like this:

<img width="609" alt="registry-3" src="https://user-images.githubusercontent.com/1181799/129103525-86493621-0544-4eb3-8a23-65094d4f16af.png">

The Custom handler decides which one to call based on the privileges and skipping the Basic Auth if it is a read operation. Yeap, I like it